Lighting/Src/lighting_controller.cpp

```diff
@@ -0,0 +1,348 @@
+	// set the index in the domain
+	this->domain_leds[domain] |= 1 << index;
+}
+
+void LightingController::remove_led_from_cd(uint8_t index,
+		ControlDomain domain) {
+	// clear the index from the domain
+	this->domain_leds[domain] &= ~(1 << index);
+}
+
+void LightingController::set_domain_colour(ControlDomain domain,
+		RGB_colour_t colour, uint8_t brightness) {
+	this->domain_colours[domain] = colour;
+	this->domain_brightness[domain] = brightness;
+}
+
+void LightingController::enable_control_domain(ControlDomain domain) {
+	this->domain_state[domain] = true;	// TODO: have active/disabled booleans
+	for (int i = 0; i < CD_LENGTH; ++i) {
+		if (this->domain_state[i]) { // IF THIS DOMAIN IS ACTIVE
+			for (int j = 0; j < NUM_LEDS; ++j) {
+				if (this->domain_leds[i] & (1 << j)) {
+					this->leds[j].set_led_colour(domain_colours[i],
+							domain_brightness[i]);
+				}
+			}
+		}
+	}
+}
+
+void LightingController::disable_control_domain(ControlDomain domain) {
+	this->domain_state[domain] = false;
+	for (int i = 0; i < CD_LENGTH; ++i) {
+		if (this->domain_state[domain]) { // IF THIS DOMAIN IS ACTIVE
+			for (int j = 0; j < NUM_LEDS; ++j) {
+				if (this->domain_leds[i] & (1 << j)) {
+					this->leds[i].set_led_colour(domain_colours[i],
+							domain_brightness[i]);
+				}
+			}
+		}
+	}
+}
+
+/////////////////
+// Private fn
+/////////////////
+void LightingController::initialize_bank_buffer_off() {
+	for (int i = 0; i < BANK_OUTPUT_BUFFER_SIZE; ++i) {
+		// Check if the bit is a padding bit or value bit
+		if (i < 24 * NUM_LEDS_PADDING
+				|| i >= (BANK_OUTPUT_BUFFER_SIZE - 24 * NUM_LEDS_PADDING)) {
+			this->bank_buffer[i] = 0;
+		} else {
+			this->bank_buffer[i] = PWM_LO;
+		}
+	}
+}
+
+void LightingController::initialize_bank_buffer_on() {
+	for (int i = 0; i < BANK_OUTPUT_BUFFER_SIZE; ++i) {
+		// Check if the bit is a padding bit or value bit
+		if (i < 24 * NUM_LEDS_PADDING
+				|| i >= (BANK_OUTPUT_BUFFER_SIZE - 24 * NUM_LEDS_PADDING)) {
+			this->bank_buffer[i] = 0;
+		} else {
+			if ((i % 8) > 4) {
+				this->bank_buffer[i] = PWM_HI;
+			} else {
+				this->bank_buffer[i] = PWM_LO;
+			}
+		}
+	}
+}
+
+void LightingController::initialize_dma_buffer() {
+	// memcpy first bank
+	std::memcpy(this->dma_buffer, bank_output_buffer, BANK_OUTPUT_BUFFER_SIZE);
+	// memcpy the second bank
+	std::memcpy(this->dma_buffer + BANK_OUTPUT_BUFFER_SIZE, bank_output_buffer,
+			BANK_OUTPUT_BUFFER_SIZE);
+}
+
+//////////////////////////////////////////
+// CALLBACKS
+//////////////////////////////////////////
+// TODO: Register custom callbacks for the timers
+// https://community.st.com/t5/stm32-mcus/how-to-use-register-callbacks-in-stm32/ta-p/580499
+// We should be able to register _class_ functions as callbacks?
+void HAL_TIM_PWM_PulseFinishedHalfCpltCallback(TIM_HandleTypeDef *htim) {
+	// | BANK 1 | BANK 2 |
+	//          ^ Current location
+	// So update BANK 1
+	std::memcpy(dma_output_buffer, bank_output_buffer, BANK_OUTPUT_BUFFER_SIZE);
```

Comment:
I guess if you really want to make this look nice. You can encapsulate this callback handle into a method and call `rev3.update_DMA_buffer(bank_offset)` in the HAL interrupt callback. 
This is similar to what you have done in `TIM7_100msCallback`. The benefit will be `dma_output_buffer` and `bank_output_buffer` can be created in a class instead of global. And the `rev3` is already global, so you don't lose anything.